docs/sphinx/specification/cudaq/types.rst

```diff
@@ -4,30 +4,91 @@ The CUDA Quantum language extension provides certain fundamental types that are
 
-**[6]** All instantiated :code:`cudaq::qudit` instances start in the :code:`|0>` computational basis 
-state and serve as the primary input argument for quantum intrinsic operations 
-modeling typical logical quantum gate operations.
+**[2]** CUDA Quantum kernels shall not have objects of type :code:`qudit` with
+        different dimensions.
 
-**[7]** :code:`cudaq::qudit` instances can be negated when leveraged as controls in 
-quantum operation invocations. The mechanism for negation should be via overloading of :code:`qudit<N>::operator!()`.
+**[3]** Objects of type :code:`qudit` shall not carry a quantum state.
+
+**[4]** Objects of type :code:`qudit` shall identify a qudit in the underlying
+        quantum system. The identifier shall be unique and valid during the
+        lifetime of the :code:`qudit` object.
+
+**[5]** Objects of type :code:`qudit` and any directly or indirectly derived
+        type shall only be used within quantum kernel code.
+
+**[6]** The default constructor shall create an object of type code:`qudit` that
+        references a qudit in the :math:`|0\rangle` computational basis state.
+
+**[7]** Objects of type :code:`qudit` may be non-copyable.
+
+**[8]** Objects of type :code:`qudit` may be non-movable.
+
+**[9]** Objects of type :code:`qudit` may have a mechanism that indicates the
+        presence of implicit Pauli-X operators surrounding the qudit use as a
+        control. For example (pseudocode):
+        .. code-block:: cpp
+          cudaq::qudit control;
+          cudaq::qudit target;
+
+          op(!control, target) // Implicates: x(control)
+                               //             op(control, target)
+                               //             x(control)
+
+C++ :code:`cudaq::qudit<Levels>`
+--------------------------------
+
+The C++ implementation of CUDA Quantum defines the templated type
+:code:`cudaq::qudit` to references qudits. The template parameter defines
+the dimension, :math:`d`, of the qudit, also known as the number of logical
+levels allowed by the qudit (hence use of `Levels` as parameter name).
+
+The :code:`cudaq::qudit` type encapsulates a unique identifier (of type 
+:code:`std::size_t`) modeling the index of the qudit in the underlying quantum
+memory space (assuming an infinite register of available qudits). This
+identifier is unique and valid only through the lifetime of the object.
+
+To facilitate memory handling, by hiding from the developer its details,
+instances of :code:`cudaq::qudit` are non-copyable and non-movable. Therefore,
+all objects of this type must be passed by reference. Qudits are implicitly
+deallocated upon the destruction of `cudaq::qudit` object that references it.
+
+:code:`cudaq::qudit` instances can be negated when leveraged as controls in 
+quantum operators. The mechanism for negation is via overloading of
+:code:`qudit<Levels>::operator!()`.
+
+In library mode, it is the runtime responsibilities to:
```

Comment:
I see library-mode + MLIR mode as implementation details, and not something we need for this document.